src/ArraySet/ArraySet.ts

```diff
@@ -0,0 +1,203 @@
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of(['A', 'B']),
+ *   ArraySet.add('C')
+ * ) // => ['A', 'B', 'C']
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of(['A', 'B']),
+ *   ArraySet.add('A')
+ * ) // => ['A', 'B']
+ */
+export const add =
+    <T>(elem: NoInfer<T>, equalityComparer?: EqualityComparer<NoInfer<T>>) =>
+    (set: ArraySet<T>): ArraySet<T> =>
+        pipe(
+            set,
+            Array.contains(elem, equalityComparer),
+            Bool.match({
+                true: set,
+                false: unsafe([elem, ...set]),
+            })
+        )
+
+/**
+ * Remove an element from the `ArraySet`, using the given `EqualityComparer` if provided.
+ * If the element does not exist, will return an `ArraySet` with the same elements.
+ *
+ * @group Transformations
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of(['A', 'B', 'C']),
+ *   ArraySet.remove('B')
+ * ) // => ['A', 'C']
+ */
+export const remove =
+    <T>(
+        elem: NoInfer<T>,
+        { equals }: EqualityComparer<NoInfer<T>> = EqualityComparer.Default
+    ) =>
+    (set: ArraySet<T>): ArraySet<T> =>
+        pipe(
+            set,
+            Array.filter(el => !equals(el, elem)),
+            unsafe
+        )
+
+/**
+ * Convert an existing `Array` into an `ArraySet`, using the given `EqualityComparer`.
+ * Analogous to `new Set([...])`.
+ *
+ * @group Constructors
+ */
+export const ofArray = <T>(
+    arr: readonly T[],
+    equalityComparer?: EqualityComparer<T>
+): ArraySet<T> => pipe(arr, Array.uniq(equalityComparer), unsafe)
+
+/**
+ * Alias for {@link ofArray}
+ *
+ * @group Constructors
+ */
+export const of = ofArray
+
+/**
+ * Create a new `ArraySet` from the result of merging two `ArraySet`s and keeping only
+ * the unique values, according to the given `EqualityComparer` (or default/triple-equals
+ * if not specified).
+ *
+ * @group Set Arithmetic
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of([1, 2, 3, 4]),
+ *   ArraySet.union(ArraySet.of([3, 4, 5]))
+ * ) // => [1, 2, 3, 4, 5]
+ */
+export const union =
+    <T>(set2: ArraySet<T>, equalityComparer?: EqualityComparer<T>) =>
+    (set1: ArraySet<T>): ArraySet<T> =>
+        pipe([...set1, ...set2], Array.uniq(equalityComparer), unsafe)
+
+/**
+ * Create a new `ArraySet` by keeping only elements that are contained in both `ArraySet`s.
+ * Uses the given `EqualityComparer` if specified; defaults to reference/triple-equals.
+ *
+ * @group Set Arithmetic
+ *
+ * @example
+ * pipe(
+ *   ArraySet.of([1, 2, 3, 4]),
+ *   ArraySet.intersect(ArraySet.of([3, 4, 5]))
+ * ) // => [3, 4]
+ */
+export const intersect =
+    <T>(set2: ArraySet<T>, equalityComparer?: EqualityComparer<T>) =>
+    (set1: ArraySet<T>): ArraySet<T> => {
+        const sets = [set1, set2] as const
+
+        const intersection: T[] = []
+
+        const smallerIdx = set1.length > set2.length ? 1 : 0
+        const biggerIdx = smallerIdx === 1 ? 0 : 1
+
+        const smaller = sets[smallerIdx]
+        const bigger = sets[biggerIdx]
```

Comment:
⛏️ Feels like this could be simpler. 🤔 
Something like:
```ts
const smaller = set1.length < set2.length ? set1 : set2
const bigger = smaller === set1 ? set2 : set1 // Or check set1.length >= set2.length
```

Reply:
ah. yep, for some reason I just totally forgot that I have reference equality checks available 🤦‍♂️ will fix

Reply:
https://github.com/fp-toolkit/fp-toolkit/pull/67/commits/f44500c18ba97d3e1a597457e851ad69cce80ff6